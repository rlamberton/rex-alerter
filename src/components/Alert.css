.alert {
    list-style-type: none;
    color: white;
    margin: 16px;
    padding: 4px;
    border-radius: 10px;
    max-width: 1200px;
    font-size: 32px;
  }
  
  .alert a {
    display: flex;
    justify-content: space-around;
    align-items: center;
    color: white;
    text-decoration: none;
  }
  
  .alert div {
    width: -webkit-fill-available;
  }
  
  .alert img {
    height: 50px;
   
  }
  
  .alert.veryLargeDrop {
    background-color: red;
    font-size: 50px;
    font-weight: bold;
  }
  
  .alert.largeDrop {
    background-color: red;
  }
  
  .alert.mediumDrop {
    background-color: purple;
  }
  
  .alert.smallDrop {
    background-color: orange;
  }
  
  .alert.increase {
    background-color: green;
  }
  
  .alert.largeIncrease {
    background-color: blue;
  }
  
  .alert.nochange {
    background-color: grey;
  }
  
  .alert .smallFont {
    font-size: 18px;
  }
  
  .rotateLeft90{
    transform: rotate(-90deg);
  }
  
  .rotateRight90{
    transform: rotate(90deg);
  }