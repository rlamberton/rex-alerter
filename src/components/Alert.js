
/**
 * React functional component for each Alert
 */
function Alert(props) {
    const alert = props.item;
    const classname = 'alert ' + getColour(alert);
    const arrowClass = (alert.pctChange < 0) ? 'rotateRight90' : 'rotateLeft90';

    return  <li className={classname}>
                <a href={alert.url} target="_blank">
                    <div>{alert.time}</div>
                    <div>{alert.symbol}</div>
                    <div className='smallFont'>
                        {alert.oldPrice}
                        <div className={arrowClass}> &#x279C; </div>
                        {alert.newPrice}
                    </div>
                    <div>{alert.pctChange}%</div>
                </a>
            </li>;
};

/**
 * Determine the colour for the component, depending on the percentage change in price
 */
function getColour(alert) {
    if (alert.pctChange < -20) {
        return 'red';
    } else if (alert.pctChange < -10) {
        return 'pink';
    } else if (alert.pctChange < 0) {
        return 'orange';
    } else if (alert.pctChange == 0) {
        return 'grey';
    } else {
        return 'green';
    }
}

export default Alert;
