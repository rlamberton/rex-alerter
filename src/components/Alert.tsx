import React from 'react';
import AlertPropsType from '../types/AlertPropsType';
import AlertType from '../types/AlertType';
import './Alert.css';

const BEEP_ENABLED : boolean = false;     // Disabled as it was annoying

/**
 * React functional component for each Alert
 */
const alert : React.FC<AlertPropsType> = function Alert(props) {
    const alert : AlertType = props.item;
    const classname : string = 'alert ' + getClassName(alert.pctChange);
    const arrowClass: string = (alert.pctChange < 0) ? 'rotateRight90' : 'rotateLeft90';

    return  <li className={classname}>
                <a href={alert.url} target="_blank">
                    <div className='smallFont'>{alert.time}</div>
                    <img src={alert.logoUrl}/>
                    <div>
                        <div>{alert.symbol}</div>
                        <div className='smallFont'>{alert.name}</div>
                    </div>
                    <div className='smallFont'>vol: &#8383;{alert.volume}</div>
                    <div className='smallFont'>
                        {alert.oldPrice}
                        <div className={arrowClass}> &#x279C; </div>
                        {alert.newPrice}
                    </div>
                    <div>{alert.pctChange}%</div>
                </a>
            </li>;
}

/**
 * Determine the colour for the component, depending on the percentage change in price
 */
function getClassName(pctChange : number) : string {
    if (pctChange < -20) {
        beep();
        return 'veryLargeDrop';
    } else if (pctChange < -10) {
        beep();
        return 'largeDrop';
    } else if (pctChange < -6) {
        return 'mediumDrop';
    } else if (pctChange < -3) {
        return 'smallDrop';
    } else if (pctChange == 0) {
        return 'nochange';
    } else if (pctChange > 10) {
        return 'largeIncrease';
    } else {
        return 'increase';
    }
}

/**
 * Play a short beep sound
 */
function beep() : void {
    if (BEEP_ENABLED) {
        var snd = new Audio("data:audio/wav;base64,//uQRAAAAWMSLwUIYAAsYkXgoQwAEaYLWfkWgAI0wWs/ItAAAGDgYtAgAyN+QWaAAihwMWm4G8QQRDiMcCBcH3Cc+CDv/7xA4Tvh9Rz/y8QADBwMWgQAZG/ILNAARQ4GLTcDeIIIhxGOBAuD7hOfBB3/94gcJ3w+o5/5eIAIAAAVwWgQAVQ2ORaIQwEMAJiDg95G4nQL7mQVWI6GwRcfsZAcsKkJvxgxEjzFUgfHoSQ9Qq7KNwqHwuB13MA4a1q/DmBrHgPcmjiGoh//EwC5nGPEmS4RcfkVKOhJf+WOgoxJclFz3kgn//dBA+ya1GhurNn8zb//9NNutNuhz31f////9vt///z+IdAEAAAK4LQIAKobHItEIYCGAExBwe8jcToF9zIKrEdDYIuP2MgOWFSE34wYiR5iqQPj0JIeoVdlG4VD4XA67mAcNa1fhzA1jwHuTRxDUQ//iYBczjHiTJcIuPyKlHQkv/LHQUYkuSi57yQT//uggfZNajQ3Vmz+Zt//+mm3Wm3Q576v////+32///5/EOgAAADVghQAAAAA//uQZAUAB1WI0PZugAAAAAoQwAAAEk3nRd2qAAAAACiDgAAAAAAABCqEEQRLCgwpBGMlJkIz8jKhGvj4k6jzRnqasNKIeoh5gI7BJaC1A1AoNBjJgbyApVS4IDlZgDU5WUAxEKDNmmALHzZp0Fkz1FMTmGFl1FMEyodIavcCAUHDWrKAIA4aa2oCgILEBupZgHvAhEBcZ6joQBxS76AgccrFlczBvKLC0QI2cBoCFvfTDAo7eoOQInqDPBtvrDEZBNYN5xwNwxQRfw8ZQ5wQVLvO8OYU+mHvFLlDh05Mdg7BT6YrRPpCBznMB2r//xKJjyyOh+cImr2/4doscwD6neZjuZR4AgAABYAAAABy1xcdQtxYBYYZdifkUDgzzXaXn98Z0oi9ILU5mBjFANmRwlVJ3/6jYDAmxaiDG3/6xjQQCCKkRb/6kg/wW+kSJ5//rLobkLSiKmqP/0ikJuDaSaSf/6JiLYLEYnW/+kXg1WRVJL/9EmQ1YZIsv/6Qzwy5qk7/+tEU0nkls3/zIUMPKNX/6yZLf+kFgAfgGyLFAUwY//uQZAUABcd5UiNPVXAAAApAAAAAE0VZQKw9ISAAACgAAAAAVQIygIElVrFkBS+Jhi+EAuu+lKAkYUEIsmEAEoMeDmCETMvfSHTGkF5RWH7kz/ESHWPAq/kcCRhqBtMdokPdM7vil7RG98A2sc7zO6ZvTdM7pmOUAZTnJW+NXxqmd41dqJ6mLTXxrPpnV8avaIf5SvL7pndPvPpndJR9Kuu8fePvuiuhorgWjp7Mf/PRjxcFCPDkW31srioCExivv9lcwKEaHsf/7ow2Fl1T/9RkXgEhYElAoCLFtMArxwivDJJ+bR1HTKJdlEoTELCIqgEwVGSQ+hIm0NbK8WXcTEI0UPoa2NbG4y2K00JEWbZavJXkYaqo9CRHS55FcZTjKEk3NKoCYUnSQ0rWxrZbFKbKIhOKPZe1cJKzZSaQrIyULHDZmV5K4xySsDRKWOruanGtjLJXFEmwaIbDLX0hIPBUQPVFVkQkDoUNfSoDgQGKPekoxeGzA4DUvnn4bxzcZrtJyipKfPNy5w+9lnXwgqsiyHNeSVpemw4bWb9psYeq//uQZBoABQt4yMVxYAIAAAkQoAAAHvYpL5m6AAgAACXDAAAAD59jblTirQe9upFsmZbpMudy7Lz1X1DYsxOOSWpfPqNX2WqktK0DMvuGwlbNj44TleLPQ+Gsfb+GOWOKJoIrWb3cIMeeON6lz2umTqMXV8Mj30yWPpjoSa9ujK8SyeJP5y5mOW1D6hvLepeveEAEDo0mgCRClOEgANv3B9a6fikgUSu/DmAMATrGx7nng5p5iimPNZsfQLYB2sDLIkzRKZOHGAaUyDcpFBSLG9MCQALgAIgQs2YunOszLSAyQYPVC2YdGGeHD2dTdJk1pAHGAWDjnkcLKFymS3RQZTInzySoBwMG0QueC3gMsCEYxUqlrcxK6k1LQQcsmyYeQPdC2YfuGPASCBkcVMQQqpVJshui1tkXQJQV0OXGAZMXSOEEBRirXbVRQW7ugq7IM7rPWSZyDlM3IuNEkxzCOJ0ny2ThNkyRai1b6ev//3dzNGzNb//4uAvHT5sURcZCFcuKLhOFs8mLAAEAt4UWAAIABAAAAAB4qbHo0tIjVkUU//uQZAwABfSFz3ZqQAAAAAngwAAAE1HjMp2qAAAAACZDgAAAD5UkTE1UgZEUExqYynN1qZvqIOREEFmBcJQkwdxiFtw0qEOkGYfRDifBui9MQg4QAHAqWtAWHoCxu1Yf4VfWLPIM2mHDFsbQEVGwyqQoQcwnfHeIkNt9YnkiaS1oizycqJrx4KOQjahZxWbcZgztj2c49nKmkId44S71j0c8eV9yDK6uPRzx5X18eDvjvQ6yKo9ZSS6l//8elePK/Lf//IInrOF/FvDoADYAGBMGb7FtErm5MXMlmPAJQVgWta7Zx2go+8xJ0UiCb8LHHdftWyLJE0QIAIsI+UbXu67dZMjmgDGCGl1H+vpF4NSDckSIkk7Vd+sxEhBQMRU8j/12UIRhzSaUdQ+rQU5kGeFxm+hb1oh6pWWmv3uvmReDl0UnvtapVaIzo1jZbf/pD6ElLqSX+rUmOQNpJFa/r+sa4e/pBlAABoAAAAA3CUgShLdGIxsY7AUABPRrgCABdDuQ5GC7DqPQCgbbJUAoRSUj+NIEig0YfyWUho1VBBBA//uQZB4ABZx5zfMakeAAAAmwAAAAF5F3P0w9GtAAACfAAAAAwLhMDmAYWMgVEG1U0FIGCBgXBXAtfMH10000EEEEEECUBYln03TTTdNBDZopopYvrTTdNa325mImNg3TTPV9q3pmY0xoO6bv3r00y+IDGid/9aaaZTGMuj9mpu9Mpio1dXrr5HERTZSmqU36A3CumzN/9Robv/Xx4v9ijkSRSNLQhAWumap82WRSBUqXStV/YcS+XVLnSS+WLDroqArFkMEsAS+eWmrUzrO0oEmE40RlMZ5+ODIkAyKAGUwZ3mVKmcamcJnMW26MRPgUw6j+LkhyHGVGYjSUUKNpuJUQoOIAyDvEyG8S5yfK6dhZc0Tx1KI/gviKL6qvvFs1+bWtaz58uUNnryq6kt5RzOCkPWlVqVX2a/EEBUdU1KrXLf40GoiiFXK///qpoiDXrOgqDR38JB0bw7SoL+ZB9o1RCkQjQ2CBYZKd/+VJxZRRZlqSkKiws0WFxUyCwsKiMy7hUVFhIaCrNQsKkTIsLivwKKigsj8XYlwt/WKi2N4d//uQRCSAAjURNIHpMZBGYiaQPSYyAAABLAAAAAAAACWAAAAApUF/Mg+0aohSIRobBAsMlO//Kk4soosy1JSFRYWaLC4qZBYWFRGZdwqKiwkNBVmoWFSJkWFxX4FFRQWR+LsS4W/rFRb/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////VEFHAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAU291bmRib3kuZGUAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMjAwNGh0dHA6Ly93d3cuc291bmRib3kuZGUAAAAAAAAAACU=");
        snd.play();
    }
}

export default React.memo(alert);
