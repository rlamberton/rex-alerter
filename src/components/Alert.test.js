import { render } from '@testing-library/react';
import Alert from './Alert';

const data = {
    url: 'http://url.com',
    time: '10:30',
    logoUrl: 'http:/logo.com',
    symbol: 'XYZ',
    name: 'crypto-name',
    volume: 1000,
    oldPrice: 2000,
    newPrice: 2200,
    pctChange : 10
};

test('check snapshot', () => {
    const { asFragment } = render(<Alert item={data} />);
    expect( asFragment() ).toMatchSnapshot();
});

test('check percentage - veryLargeDrop', () => {
    const { getByTestId } = render(<Alert item={{...data, pctChange: -25}} />);
    const pctElement = getByTestId('alert-volume');
    const alertElement = getByTestId('alert');
    const arrowElement = getByTestId('alert-arrow');
    expect(pctElement).toBeInTheDocument();
    expect(alertElement).toHaveAttribute('class', 'alert veryLargeDrop');
    expect(arrowElement).toHaveAttribute('class', 'rotateRight90');
});

test('check percentage - largeDrop', () => {
    const { getByTestId } = render(<Alert item={{...data, pctChange: -15}} />);
    const pctElement = getByTestId('alert-volume');
    const alertElement = getByTestId('alert');
    const arrowElement = getByTestId('alert-arrow');
    expect(pctElement).toBeInTheDocument();
    expect(alertElement).toHaveAttribute('class', 'alert largeDrop');
    expect(arrowElement).toHaveAttribute('class', 'rotateRight90');
});

test('check percentage - mediumDrop', () => {
    const { getByTestId } = render(<Alert item={{...data, pctChange: -7}} />);
    const pctElement = getByTestId('alert-volume');
    const alertElement = getByTestId('alert');
    const arrowElement = getByTestId('alert-arrow');
    expect(pctElement).toBeInTheDocument();
    expect(alertElement).toHaveAttribute('class', 'alert mediumDrop');
    expect(arrowElement).toHaveAttribute('class', 'rotateRight90');
});

test('check percentage - smallDrop', () => {
    const { getByTestId } = render(<Alert item={{...data, pctChange: -4}} />);
    const pctElement = getByTestId('alert-volume');
    const alertElement = getByTestId('alert');
    const arrowElement = getByTestId('alert-arrow');
    expect(pctElement).toBeInTheDocument();
    expect(alertElement).toHaveAttribute('class', 'alert smallDrop');
    expect(arrowElement).toHaveAttribute('class', 'rotateRight90');
});

test('check percentage - nochange', () => {
    const { getByTestId } = render(<Alert item={{...data, pctChange: 0}} />);
    const pctElement = getByTestId('alert-volume');
    const alertElement = getByTestId('alert');
    const arrowElement = getByTestId('alert-arrow');
    expect(pctElement).toBeInTheDocument();
    expect(alertElement).toHaveAttribute('class', 'alert nochange');
    expect(arrowElement).toHaveAttribute('class', 'rotateLeft90');
});

test('check percentage - largeIncrease', () => {
    const { getByTestId } = render(<Alert item={{...data, pctChange: 20}} />);
    const pctElement = getByTestId('alert-volume');
    const alertElement = getByTestId('alert');
    const arrowElement = getByTestId('alert-arrow');
    expect(pctElement).toBeInTheDocument();
    expect(alertElement).toHaveAttribute('class', 'alert largeIncrease');
    expect(arrowElement).toHaveAttribute('class', 'rotateLeft90');
});

test('check percentage - increase', () => {
    const { getByTestId } = render(<Alert item={data} />);
    const pctElement = getByTestId('alert-volume');
    const alertElement = getByTestId('alert');
    const arrowElement = getByTestId('alert-arrow');
    expect(pctElement).toBeInTheDocument();
    expect(alertElement).toHaveAttribute('class', 'alert increase');
    expect(arrowElement).toHaveAttribute('class', 'rotateLeft90');
});

test('check urls', () => {
    const { getByTestId } = render(<Alert item={{...data, url: 'url', logoUrl: 'logo'}} />);
    const urlElement = getByTestId('alert-url');
    expect(urlElement).toHaveAttribute('href', 'url')
    const logoUrlElement = getByTestId('alert-logo-url');
    expect(logoUrlElement).toHaveAttribute('src', 'logo');
});
