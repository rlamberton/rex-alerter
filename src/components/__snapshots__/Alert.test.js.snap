// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check snapshot 1`] = `
<DocumentFragment>
  <li
    class="alert increase"
    data-testid="alert"
  >
    <a
      data-testid="alert-url"
      href="http://url.com"
      target="_blank"
    >
      <div
        class="smallFont"
        data-testid="alert-time"
      >
        10:30
      </div>
      <img
        data-testid="alert-logo-url"
        src="http:/logo.com"
      />
      <div>
        <div
          data-testid="alert-symbol"
        >
          XYZ
        </div>
        <div
          class="smallFont"
          data-testid="alert-name"
        >
          crypto-name
        </div>
      </div>
      <div
        class="smallFont"
        data-testid="alert-volume"
      >
        vol: ₿1000
      </div>
      <div
        class="smallFont"
        data-testid="alert-price"
      >
        2000
        <div
          class="rotateLeft90"
          data-testid="alert-arrow"
        >
           ➜ 
        </div>
        2200
      </div>
      <div
        data-testid="alert-pct-change"
      >
        10%
      </div>
    </a>
  </li>
</DocumentFragment>
`;
